### PR TITLE
Letsencrypt: add account management module

### LINKED ISSUE #36346: ACME: allow account management
##### ISSUE TYPE
Feature Idea

##### COMPONENT NAME
acme_certificate

##### SUMMARY
The `letsencrypt` module automatically creates an account and associates an email address with it, but doesn't allow any advanced account management features, such as updating contact data (https://tools.ietf.org/html/draft-ietf-acme-acme-09#section-7.3.2), changing the account key (https://tools.ietf.org/html/draft-ietf-acme-acme-09#section-7.3.6), or deactivating it (https://tools.ietf.org/html/draft-ietf-acme-acme-09#section-7.3.7). It would be nice to be able to do this with ansible as well.

I think this should also go into a new module, though. I'm still opening this as a issue for the `letsencrypt` module so it reaches all current maintainers :)

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Adds a new module which allows to manage Let's Encrypt accounts (using the ACME v2 protocol). Fixes #36346.

##### ISSUE TYPE
New Module Pull Request

##### COMPONENT NAME
letsencrypt_account

##### ANSIBLE VERSION
2.6

##### ADDITIONAL INFORMATION
There are some integration tests. I'm using an EC account key so that this will be fast. This test depends on the Let's Encrypt ACME v2 staging endpoint. If Ansible tests should not depend on external services, I will remove the tests for now.

### CODE REVIEW

**Comment by felixfontein** [comment]
(The first commit is for https://github.com/ansible/ansible/pull/36362#pullrequestreview-97493696)

**Comment by felixfontein** [comment]
Note to self: 5e02f1ca32e52356bcea76f50e279e6015e19b20 should go into Ansible 2.5 if it works.

**Comment by felixfontein** [comment]
Looks like everything works, except that on OSX signing with the EC account key doesn't work.

**Comment by felixfontein** [comment]
Looks like ECC with OSX' OpenSSL seems to be broken (https://security.stackexchange.com/a/56065).

**Comment by felixfontein** [comment]
Not doing the tests for OpenSSL 0.9.8 (or earlier) solved the problem. Next comes account key rollover.

**Comment by felixfontein** [comment]
I think this is done (modulo review/bugs/... :) ). The next step (in a new PR) would be to add settings to the `letsencrypt` module to stop it from creating accounts or modifying contact data (it only supports one email address, nothing else).

**Comment by felixfontein** [comment]
Once #37293 is merged, I'll rebase and remove commit f3c83086ac6be60a0a09f2c5433e0bdedd8ee718.

File: lib/ansible/module_utils/letsencrypt.py [hunk 1]
```diff
@@ -0,0 +1,528 @@
+# -*- coding: utf-8 -*-
+
+# (c) 2016 Michael Gruener <michael.gruener@chaosmoon.net>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Comment (resmo) [inline lib/ansible/module_utils/letsencrypt.py]: module_utils/* must have BSD license and we need to ask every author of this file if he is ok with.

File: lib/ansible/modules/web_infrastructure/letsencrypt_account.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+      - "The state of the account, to be identified by its account key."
+      - "If the state is C(absent), the account will either not exist or be
+         deactivated."
+      - "If the state is C(changed_key), the account must exist."
+    required: true
+    choices:
+    - present
+    - absent
+    - changed_key
+    version_added: "2.5"
```
Comment (resmo) [inline lib/ansible/modules/web_infrastructure/letsencrypt_account.py]: Please remove the version_added here because module is new.

**Comment by resmo** [comment]
@felixfontein I looked through the code and I wondered what is the usecase for "state=changed_key" can you explain a bit more about your thoughts?

File: lib/ansible/modules/web_infrastructure/letsencrypt_account.py (hunk 2 above)
Comment (felixfontein) [reply lib/ansible/modules/web_infrastructure/letsencrypt_account.py]: I guess that was a copy'n'paste error.

File: lib/ansible/module_utils/letsencrypt.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/module_utils/letsencrypt.py]: Hmm, that sounds annoying. I'll compile a list somewhen later...

**Comment by felixfontein** [comment]
@resmo The `state == 'changed_key'` is for changing the account key. You have to start with your current account key, and specify the new account key with `new_account_key_*`. After the module goes through, you need to use the new account key to do anything with the account (change details, or issue certificates).
I'll add some more documentation later.

**Comment by resmo** [comment]
@felixfontein some examples would definitely be helpful.

**Comment by felixfontein** [comment]
@resmo Ah, I completely forgot an example... I've added that now.

File: lib/ansible/module_utils/letsencrypt.py (hunk 1 above)
Comment (felixfontein) [reply lib/ansible/module_utils/letsencrypt.py]: Is there a procedure for this? How must the consent of every author be documented?

Comment (gundalow) [reply lib/ansible/module_utils/letsencrypt.py]: Correct

Comment (felixfontein) [reply lib/ansible/module_utils/letsencrypt.py]: @resmo @gundalow Sorry to bother you again, but: is there a procedure for this? And how must the consent of all authors be documented?

Comment (gundalow) [reply lib/ansible/module_utils/letsencrypt.py]: A comment in the PR. you can `@mention` people so you get their attention via GitHub notifications

Comment (felixfontein) [reply lib/ansible/module_utils/letsencrypt.py]: Ah, that sounds easy enough. I was afraid that something more complicated (like signed letteres, ...) was needed.

**Comment by felixfontein** [comment]
Ok, here's a list of all commits affecting `letsencrypt.py` in `devel` (also in the `ansible-modules-extra` repository). I looked at every commit whether it affects code moved into `module_utils` by this PR.

| commit title | commit URL | author | affects? |
| ------------ | ---------- | ------ | ------- |
| Add Let's Encrypt module to manage ssl certificates (#1962) | https://github.com/ansible/ansible-modules-extras/commit/bb6504c8d351ea188712c6c4976fe521ec3f189c | mgruener | yes |
| letsencrypt: fix tests failures (#2360) | https://github.com/ansible/ansible-modules-extras/commit/4874ceefa59ac85a57943ac59c90c3b23d7550f3 | resmo | yes |
| letsencrypt: update URL for agreement pdf (#2696) | https://github.com/ansible/ansible-modules-extras/commit/cfff23fa4afab9276f7a097aa37598098900fd18 | rmetzler | no |
| letsencrypt: Improve error handling (#2868) | https://github.com/ansible/ansible-modules-extras/commit/d69a6f20daadf802e987697259ce40b41684f4bc | mgruener | yes |
| Fix error in description of acme_directory (#3034) | https://github.com/ansible/ansible-modules-extras/commit/7aab9cd93b6c27a44a9dcbfa5c902b4e37e4b40c | mscherer | no |
| Add missing deps, since we use openssl to verify the date | https://github.com/ansible/ansible-modules-extras/commit/a42383adc1187f0fda744bbfcab377b7585bade0 | mscherer | no |
| Use 'path' type for file arguments | https://github.com/ansible/ansible-modules-extras/commit/555439497136c127f5146509972b9adbcf8f4442 | mscherer | yes |
| Update letsencrypt.py | https://github.com/ansible/ansible-modules-extras/commit/63e6d3a4601c191ea40ec70ce1587aa53492683d | aioue | yes |
| Update letsencrypt.py | https://github.com/ansible/ansible-modules-extras/commit/c80f12c4c9e751350e966de9d89ea7a0beb1bfc7 | aioue | no |
| Fixed wrong variable name in the let's encrypt doc (#3398) | https://github.com/ansible/ansible-modules-extras/commit/81009c5f6f603b984461a14c724ed9cccf2de68a | tristanbes | no |
| letsencrypt: Locale-independent date parsing (#3314) | https://github.com/ansible/ansible-modules-extras/commit/22a6449150ef451f14028e75e1ffc09d2a989a51 | mgedmin | no |
| Refreshed metadata for extras modules | https://github.com/ansible/ansible-modules-extras/commit/6c7d63b15c77126b4d6a8a7668545555578469c5 | abadger | no |
| Relocating extras into lib/ansible/modules/ after merge | https://github.com/ansible/ansible/commit/011ea55a8f1630842c67603ac601d4d7ef6ccef9 | jimi-c | no |
| minor spelling changes | https://github.com/ansible/ansible/commit/0b8011436dc7f842b78298848e298f2a57ee8d78 | cgar | yes |
| Fix letsencrypt remaining_days docs | https://github.com/ansible/ansible/commit/778a189a590bd1a29ee58964599ed168ce225111 | rouge8 | no |
| Fix infrequent PEP 8 issues. | https://github.com/ansible/ansible/commit/63b1e0c277fd19f651bd925386f7fcb9d1e89cfb | mattclay | yes |
| PEP 8 whitespace cleanup. (#20783) | https://github.com/ansible/ansible/commit/95789f3949b01c79c729315ff375d0e759f64970 | mattclay | no |
| PEP 8 W291 whitespace cleanup. | https://github.com/ansible/ansible/commit/d913f69ba10af797912327f553239081d17a2593 | mattclay | no |
| Update validate-modules (#20932) | https://github.com/ansible/ansible/commit/829c0b8f6284ec37e36f1454d7d3640415239cb2 | sivel | no |
| Fix invalid fields in module DOCUMENATION (#22297) | https://github.com/ansible/ansible/commit/e5b990a55a77371439c373071c22d7d70e95a534 | gundalow | no |
| New metadata 1.0 (#22587) | https://github.com/ansible/ansible/commit/eb1214baad0cbe2c4b7304caebb9ae1c7dc0d8db | abadger | no |
| fixed RETURN docs for modules (#24011) | https://github.com/ansible/ansible/commit/d1517f997abc68155f6e9970c64fa027702e11ae | Shaps | no |
| Pep8 fixes for letsencrypt module (#24144) | https://github.com/ansible/ansible/commit/fbb924ff24dbe140515b46829d537e95af027647 | Akasurde | yes |
| Fix more module returns (#24782) | https://github.com/ansible/ansible/commit/ed2f13b3dba1777341d436be7b0362840043bad4 | gundalow | no |
| Letsencrypt fix (#24906) | https://github.com/ansible/ansible/commit/16d522cf2cfcd85bce377cda26865f15f299c892 | kgottholm | yes |
| Cmp python3 compat (#25008) | https://github.com/ansible/ansible/commit/5c43bd3baec1f10968a1b971edcdd1100b821e62 | kkjang | no |
| letsencrypt: FIX CN parsing to work with OpenSSL 1.1 (#25935) | https://github.com/ansible/ansible/commit/9474f20f2df6e10734d81800b5b0f57ae5c834ca | hsoft | no |
| Remove wildcard imports | https://github.com/ansible/ansible/commit/4e6cce354eb3266d8819015c23e4d80691fa8585 | abadger | yes |
| Update metadata to 1.1 | https://github.com/ansible/ansible/commit/f203ca7907bd3fc0e1882ab65960e6720907fe0f | abadger | no |
| Fix trailing space in ec2_vol example, fix 'the the' typos (#28440) | https://github.com/ansible/ansible/commit/beed59f303e1efbf8e6b6d5f7e5fc985f883a1e4 | davidkretch | no |
| Fix incorrect use of datetime in letsencrypt (#27672) | https://github.com/ansible/ansible/commit/551a020e785fa2253e47eefe6312781e7dbaf91d | jimi-c | no |
| Fix dangerous default args. (#29839) | https://github.com/ansible/ansible/commit/68aeaa58a89fdcf31511ddc61d7e4b51c0099d6a | mattclay | yes |
| Add a code-smell test for smart quotes and remove smart quotes from a… | https://github.com/ansible/ansible/commit/c82cf791dd82e7e59268be91b885f82a5bd42a0a | abadger | no |
| Add support for generating full chain certificates from LetsEncrypt. (#… | https://github.com/ansible/ansible/commit/f71816c192c0079f51fa93287e55721374bd6ec7 | manicai | no |
| letsencrypt: fix doc version_added (#32943) | https://github.com/ansible/ansible/commit/97b844e66083c7fd620d4e77e7a79bf7c9d8af6b | resmo | no |
| letsencrypt: update agreement default to newest gathered | https://github.com/ansible/ansible/commit/c64eed16fe593cc3685b0fff3a79c9a239b884b3 | resmo | yes |
| Add python level locale handling back (#31339) | https://github.com/ansible/ansible/commit/fd4a6cf7ad7012a07b29096f5c70f34499af1216 | skazi0 | no |
| Deprecate tests used as filters (#32361) | https://github.com/ansible/ansible/commit/4fe08441be093a9cbe968e00606e03e33605207c | sivel | no |
| letsencrypt: fix docs, openssl csr to req (#33519) | https://github.com/ansible/ansible/commit/54825190960b367178075f703ba36fd2fe040aa2 | tiggi | no |
| Bulk autopep8 (modules) | https://github.com/ansible/ansible/commit/c57a7f05e1b37084110d2b1bc219b40fc9023ffa | gundalow | yes |
| python3 support for letsencrypt module (fixes #30690) (#32734) | https://github.com/ansible/ansible/commit/3a634058f3c417339170b3764f3b899224273d0c | bverschueren | yes |
| letsencrypt: add new param account_key_content (#32948) | https://github.com/ansible/ansible/commit/d307dee4db9039514f102ddc2cb2638b6e295abd | resmo | yes |
| letsencrypt: fix fullchain is False by default. (#34912) | https://github.com/ansible/ansible/commit/8d69eb44882c49e56f50e1308ad11d1d170bd4bf | felixfontein | no |
| letsencrypt: Add support for Elliptic Curve account keys (#34852) | https://github.com/ansible/ansible/commit/c42c0f0cb388b052d964ecea57d8ba79ba004045 | felixfontein | yes |
| Secure letsencrypt account_key_content better (#34979) | https://github.com/ansible/ansible/commit/1d26686b0994603ac4f6d2271f162b549e3f2060 | felixfontein | yes |
| Making letsencrypt module work with Python 3. (#34324) | https://github.com/ansible/ansible/commit/80361ce4da747547fac3b6ab823c900479117257 | felixfontein | yes |
| Clarifying documentation about account key being different from certi… | https://github.com/ansible/ansible/commit/3a4b69096f68cf73ed8fc45cbd4b8be429871170 | felixfontein | no |
| Extending documentation (#35077) | https://github.com/ansible/ansible/commit/39a7e0a9754b3e554b0f0065ea3c0a652c39ed60 | felixfontein | no |
| Accepting SANs marked as critical (fixes #32767). (#35057) | https://github.com/ansible/ansible/commit/784853da35762ffd82742fb11e67001dbd3d0edc | felixfontein | no |
| tempfile.mkstemp() actually returns a file descriptor which should be… | https://github.com/ansible/ansible/commit/0f17d74f0faa90d66ed1e39a34d9310a5051b105 | felixfontein | yes |
| Changing letsencrypt module to use ACME v2 protocol (#34541) | https://github.com/ansible/ansible/commit/8095815b3299f6acda3b285639346980374c4162 | felixfontein | yes |
| Removing underscore variables in letsencrypt module (#35282) | https://github.com/ansible/ansible/commit/5959b93248ad729e703e165cb7375c3dab1c6b1e | felixfontein | yes |
| Letsencrypt: cleaning up tempfile code (2) (#35278) | https://github.com/ansible/ansible/commit/b50ab8eebd5b298092ddcaa3034a2f4663bedb85 | felixfontein | yes |
| Changing behavior of new fullchain argument. (#34916) | https://github.com/ansible/ansible/commit/e75055923b0f9c7b4c56fe678ef8c3d5b9194b7f | felixfontein | no |
| Allowing to write intermediate certificate into file. (#34328) | https://github.com/ansible/ansible/commit/a69445e2189684c0277565e12b88e20caaec4ce5 | felixfontein | no |
| Changing letsencrypt module to use ACME v2 protocol (update) (#35283) | https://github.com/ansible/ansible/commit/2ebb611b502d56cb91f7e33314e827f580be1ada | felixfontein | yes |
| letsencrypt: fix account key detection for keys created by openssl_pr… | https://github.com/ansible/ansible/commit/2c482847ce4259283f08c1397cee4fc957075ac2 | felixfontein | yes |
| Letsencrypt authz deactivation (#36362) | https://github.com/ansible/ansible/commit/cd9d55418670dee73ac6f7a1ee0a31598c31bcd9 | felixfontein | yes |
| letsencrypt: add new param force (#37077) | https://github.com/ansible/ansible/commit/4705edd1a79cc1d55d1c1b900e1208aa7e5554f7 | resmo | no |
| Using correct content type (as per version 10 of ACME draft). (#37165) | https://github.com/ansible/ansible/commit/155adb16319236d1d891556717b34ad0db4a1879 | felixfontein | yes |
| Improving EC account key parsing (see #37275). (#37293) | https://github.com/ansible/ansible/commit/3ce325e35e8de19339a882f731231eac93f92919 | felixfontein | yes |
| Production ACME v2 endpoint is available. (#37369) | https://github.com/ansible/ansible/commit/042ad70da05e7ddeda13d93565cfd278a722a3ce | felixfontein | no |
| Clean up module documentation (#36909) | https://github.com/ansible/ansible/commit/cdd21e2170b9a47d84624c8c2f792a66f2b0fe57 | dagwieers | no |
| Fixing domain to authorization matching for ACME v2. (#37558) | https://github.com/ansible/ansible/commit/190755ff659aa18a4a05951c04c41808358121c9 | felixfontein | no |
| letsencrypt: don't mask values specified in data (#37556) | https://github.com/ansible/ansible/commit/60fee5154b15b5f7b2558d1eb65e1b3f153d9f71 | felixfontein | no |
| Remove the deprecated tls-sni-02 challenge method. (#38137) | https://github.com/ansible/ansible/commit/8b98c29afdfc299e8b03dab9d152a957e3a39153 | felixfontein | no |
| Making clearer that the module has to be executed twice, and that the… | https://github.com/ansible/ansible/commit/93fc99c5768737a74df4fc57fc97a4c1a2de072c | felixfontein | no |

**List of all authors:** abadger, aioue, Akasurde, bverschueren, cgar, dagwieers, davidkretch, felixfontein, gundalow, hsoft, jimi-c, kgottholm, kkjang, manicai, mattclay, mgedmin, mgruener, mscherer, resmo, rmetzler, rouge8, Shaps, sivel, skazi0, tiggi, tristanbes

**List of authors of affected commits:** abadger, aioue, Akasurde, bverschueren, cgar, felixfontein, gundalow, kgottholm, mattclay, mgruener, resmo

**Comment by felixfontein** [comment]
@abadger @aioue @Akasurde @bverschueren @cgar @gundalow @kgottholm @mattclay @mgruener @resmo

Dear all,
I want to add two more Let's Encrypt modules to Ansible (one for account management #36346, implemented in this PR, and one for revoking certificates #36345). Since some of the code from the `letsencrypt` module can be re-used for these tasks, I wanted to move parts of it to a Python module in the `module_utils` folder. This is done in this PR; you can see the [new Python module `module_utils/letsencrypt.py` here](https://github.com/felixfontein/ansible/blob/72c32fa84ff80b6a5bbba12e4f9fabe86194360e/lib/ansible/module_utils/letsencrypt.py).
Now all Python modules in `module_utils` must have a BSD license; the original module is licensed under GPLv3. Since you all authored commits which modified code or comments which will be moved into `module_utils` by this commit, you all need to confirm that you are OK with this license change. All code not moved to `module_utils` is not affected by this. In this PR, some of the documentation is moved to [`utils/module_docs_fragments/letsencrypt.py`](https://github.com/felixfontein/ansible/blob/72c32fa84ff80b6a5bbba12e4f9fabe86194360e/lib/ansible/utils/module_docs_fragments/letsencrypt.py); this will stay GPL licensed.
More precisely, I want to change the lines:
``` .py
# (c) 2016 Michael Gruener <michael.gruener@chaosmoon.net>
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
of `module_utils/letsencrypt.py` into
``` .py
# This code is part of Ansible, but is an independent component.
# This particular file snippet, and this file snippet only, is BSD licensed.
# Modules you write using this snippet, which is embedded dynamically by Ansible
# still belong to the author of the module, and may assign their own license
# to the complete work.
#
# Copyright (c), Michael Gruener <michael.gruener@chaosmoon.net>, 2016
#
# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
```
(Taken from [`module_utils/urls.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py).)
Please confirm by writing a comment in this PR whether you are OK with this change or not. Thank you very much in advance!

**Comment by abadger** [comment]
I give permission for any code that I contributed to ./lib/ansible/modules/web_infrastructure/letsencrypt.py to be relicensed from the GPLv3+ to the simplified BSD license mentioned in https://github.com/ansible/ansible/pull/37275#issuecomment-379469158 for the purpose of creating a letsencrypt module_util file which can be shared amongst several modules.  I understand that once the code is in the module_util file under the BSD license, it may be used in other works pursuant to the BSD license.

**Comment by Akasurde** [comment]
I give permission for any code that I contributed to be relicensed from the GPLv3+ to the simplified BSD license.

**Comment by felixfontein** [comment]
@abadger @Akasurde thanks!

**Comment by cgar** [comment]
Confirmed, I'm OK with the license change.

**Comment by bverschueren** [comment]
I give permission for any code that I contributed to ./lib/ansible/modules/web_infrastructure/letsencrypt.py to be relicensed from the GPLv3+ to the simplified BSD license.

**Comment by felixfontein** [comment]
@cgar @bverschueren thanks as well!

**Comment by mattclay** [comment]
I give permission for any code that I contributed to lib/ansible/modules/web_infrastructure/letsencrypt.py to be relicensed from the GPLv3+ to the simplified BSD license mentioned in https://github.com/ansible/ansible/pull/37275#issuecomment-379469158 for the purpose of creating a letsencrypt module_util file which can be shared amongst several modules.  I understand that once the code is in the module_util file under the BSD license, it may be used in other works pursuant to the BSD license.

**Comment by felixfontein** [comment]
@mattclay thanks!

**Comment by felixfontein** [comment]
I'm not sure whether I have to do this myself, but just in case: I give permission for any code that I contributed to be relicensed from the GPLv3+ to the simplified BSD license.

**Comment by felixfontein** [comment]
So far, feedback is missing from @aioue, @gundalow, @kgottholm, @mgruener and @resmo.

**Comment by resmo** [comment]
I give permission for any code that I contributed to be re-licensed from the GPLv3+ to the simplified BSD license.

**Comment by gundalow** [comment]
I give permission for any code that I contributed to lib/ansible/modules/web_infrastructure/letsencrypt.py to be relicensed from the GPLv3+ to the simplified BSD license mentioned in https://github.com/ansible/ansible/pull/37275#issuecomment-379469158 for the purpose of creating a letsencrypt module_util file which can be shared amongst several modules.  I understand that once the code is in the module_util file under the BSD license, it may be used in other works pursuant to the BSD license.

**Comment by felixfontein** [comment]
@resmo @gundalow thanks!

**Comment by mgruener** [comment]
I give permission for any code that I contributed to lib/ansible/modules/web_infrastructure/letsencrypt.py to be relicensed from the GPLv3+ to the simplified BSD license mentioned in #37275 (comment) for the purpose of creating a letsencrypt module_util file which can be shared amongst several modules. I understand that once the code is in the module_util file under the BSD license, it may be used in other works pursuant to the BSD license.

**Comment by felixfontein** [comment]
@mgruener thank you very much! :)

**Comment by felixfontein** [comment]
@aioue @kgottholm is any one of you still around?

**Comment by felixfontein** [comment]
@ all: since we renamed the `letsencrypt` module to `acme_certificate` (#39737, to avoid violating Let's Encrypt's trademark policy), I'll rebase this PR to current master. While doing this, I will also adjust the name of the new module to `acme_account`. I will currently *not* change the names of the `module_utils` module and the documentation fragment (or change anything else); I think it will be better to do this in a separate PR (when this one is merged) to avoid any ambiguities arising in that some of you (including me) explicitly mentioned the `letsencrypt` `module_utils` module. I hope this is OK for everyone.

**Comment by felixfontein** [comment]
@resmo @gundalow I have a question about integration tests: do I understand it correctly (from looking at some builds at shippable.com) that only integration tests belonging to the affected modules (of a PR/commit) are run, as well as their dependencies? Is it guaranteed that all other integration tests will not be ran?
I'm asking because the integration tests I added in this PR do communicate with the Let's Encrypt staging servers, and thus a) tests can fail because of network problems, and b) if these tests are run every time an ansible build runs on shippable.com, this might put some unwanted strain on the Let's Encrypt staging servers.

**Comment by felixfontein** [comment]
Ok, I was wrong, apparently some of the builds in one shippable.com run do run all integration tests. Any ideas what we should do with the tests?

**Comment by mattclay** [comment]
@felixfontein The integration tests executed by CI are dependent on the changes made in a PR. Changes to the module or its tests will trigger the tests. Changes to most non-module code/tests will trigger a more complete run of the tests, which would include the tests in this PR.

I'd keep the tests and we can revisit if there are any issues.

**Comment by felixfontein** [comment]
@mattclay Thanks for the explanation! Then let's keep the tests :)

**Comment by felixfontein** [comment]
I want to give a summary on the current state of the license change.

So far, most commit authors whose commits are affected by the license change have given their permission for the license change to take place (see below for a table with links to the comments). There has been no feedback from @aioue (the affected commit is ansible/ansible-modules-extras@63e6d3a4601c191ea40ec70ce1587aa53492683d, only its lower part affects the code which is moved into `module_utils`) and @kgottholm (the affected commit is ansible/ansible@16d522cf2cfcd85bce377cda26865f15f299c892). I've tried to contact both of them via email last Thursday (using email addresses I found in the git log or in WHOIS), with no reaction so far.

Author | Comment with permission
------ | -----------------------
abadger | https://github.com/ansible/ansible/pull/37275#issuecomment-379480479
Akasurde | https://github.com/ansible/ansible/pull/37275#issuecomment-379484234
bverschueren | https://github.com/ansible/ansible/pull/37275#issuecomment-379568807
cgar | https://github.com/ansible/ansible/pull/37275#issuecomment-379557116
felixfontein | https://github.com/ansible/ansible/pull/37275#issuecomment-381405458
gundalow | https://github.com/ansible/ansible/pull/37275#issuecomment-381574434
mattclay | https://github.com/ansible/ansible/pull/37275#issuecomment-379788145
mgruener | https://github.com/ansible/ansible/pull/37275#issuecomment-384158066
resmo | https://github.com/ansible/ansible/pull/37275#issuecomment-381411120

**Comment by felixfontein** [comment]
Here are the follow-up commits I plan after this PR is merged: https://github.com/felixfontein/ansible/compare/letsencrypt-account-management...felixfontein:letsencrypt-account-management-followup
This will rename the module utils package and documentation fragment, and add a new option to `acme_certificate` to stop it from doing account management.

**Comment by felixfontein** [comment]
I've add two commits which change the code modified by the two commits we got no reply for so far so that hopefully permission for these two commits is no longer needed:
1. The code changed in 16d522cf2cfcd85bce377cda26865f15f299c892 has been rewritten (and simplified) in f227f8c35d31fd4deb27dac3b76f350cd4e08a29;
2. The docstring updated in the lower part of ansible/ansible-modules-extras@63e6d3a4601c191ea40ec70ce1587aa53492683d (the upper part touches the `acme_certificate` documentation which isn't moved to the module_utils module) has been rewritten in 207ef043b7c136ef632466dc293d2bbb845197a9.

@resmo @gundalow Is this sufficient so that this PR can be merged?

**Review by gundalow — APPROVED** [review]

**Comment by felixfontein** [comment]
As described [here](https://github.com/ansible/ansible/pull/37275#issuecomment-379469158), I changed the license declaration for `lib/ansible/module_utils/letsencrypt.py` in fdd68f36eb22dd722b869a181f0266cc23385fc9.

**Comment by felixfontein** [comment]
I've updated BOTMETA.yml; @resmo: are you OK with [this](https://github.com/ansible/ansible/pull/37275/commits/4786cea7f7b6463780158a7e067bcf30ea482d8f)?

**Comment by felixfontein** [comment]
@resmo what do you think?

**Comment by maxamillion** [comment]
reping @resmo

**Comment by resmo** [comment]
rebuild_merge

**Review by resmo — APPROVED** [review]
shipit

**Comment by felixfontein** [comment]
bot_status

**Comment by felixfontein** [comment]
Thank you very much to all of you for your contributions!

@resmo: I've created a follow-up PR in #40697 to rename the new module_utils module and the documentation fragment.

**Pull request merged**